drivers/mcp2515/mcp2515.c

```diff
@@ -0,0 +1,412 @@
+        framebuf.data[i] = frame->data[i];
+    }
+
+    //TODO: for speedup, remove this check
+    mcp2515_spi_read(dev, TX_CTRL(mailbox), &ctrl, 1);
+    if (ctrl & MCP2515_TXBCTRL_TXREQ) {
+        DEBUG("Mailbox in use, TXB%dCTRL: 0x%02x\n", mailbox, ctrl);
+        return -1;
+    }
+
+    if (framebuf.can_id > CAN_SFF_MASK) {
+        framebuf.can_id |= CAN_EFF_FLAG;
+    }
+
+    if ((framebuf.can_id & CAN_EFF_FLAG) == CAN_EFF_FLAG) {
+        fill_extended_id(framebuf.can_id, outbuf);
+    }
+    else {
+        fill_standard_id(framebuf.can_id, outbuf);
+    }
+
+    outbuf[4] = framebuf.can_dlc;
+    memcpy(&outbuf[CAN_FIXED_LEN], framebuf.data, framebuf.can_dlc);
+
+    /* set mailbox priority */
+    mcp2515_spi_write(dev, TX_CTRL(mailbox), &prio, 1);
+
+    mcp2515_spi_write_txbuf(dev, mailbox, outbuf,
+                            CAN_FIXED_LEN + framebuf.can_dlc);
+    mcp2515_enable_irq(dev, MCP2515_CANINTE_TX0IE << mailbox);
+    mcp2515_spi_rts(dev, mailbox);
+
+    return mailbox;
+}
+
+int mcp2515_receive(candev_mcp2515_t *dev, struct can_frame *frame, int mailbox)
+{
+    uint8_t inbuf[BUFFER_SIZE];
+
+    mcp2515_spi_read_rxbuf(dev, mailbox, inbuf, BUFFER_SIZE);
+
+    /* extended id */
+    if (inbuf[1] & MCP2515_RX_IDE) {
+        frame->can_id = ((uint32_t)inbuf[0] << 21) +
+                        (((uint32_t)inbuf[1] & 0xE0) << 13) +
+                        (((uint32_t)inbuf[1] & 0x03) << 16) +
+                        ((uint32_t)inbuf[2] << 8) +
+                        inbuf[3];
+        frame->can_id |= CAN_EFF_FLAG;
+    }
+
+    /* standard id */
+    else {
+        frame->can_id = ((uint32_t)inbuf[0] << 3) +
+                        (((uint32_t)inbuf[1] & 0xE0) >> 5);
+    }
+
+    frame->can_dlc = inbuf[4];
+    memcpy(frame->data, inbuf + 5, frame->can_dlc);
+    return mailbox;
+}
+
+int mcp2515_abort(candev_mcp2515_t *dev, int mailbox)
+{
+    return mcp2515_spi_bitmod(dev, TX_CTRL(mailbox), MCP2515_TXBCTRL_TXREQ, 0);
+}
+
+enum mcp2515_mode mcp2515_get_mode(candev_mcp2515_t *dev)
+{
+    enum mcp2515_mode result = MODE_UNKNOWN;
+    uint8_t mode;
+
+    int res = mcp2515_spi_read(dev, MCP2515_CANSTAT, &mode, 1);
+
+    if (res == 0) {
+        result = (enum mcp2515_mode)mode & MCP2515_CANSTAT_OPMOD_MASK;
+    }
+
+    DEBUG("mcp2515_get_mode: mode=%x\n", result);
+
+    return result;
+}
+
+enum mcp2515_mode mcp2515_set_mode(candev_mcp2515_t *dev,
+                                   enum mcp2515_mode mode)
+{
+    DEBUG("mcp2515_set_mode: mode=%x\n", mode);
+
+    if (mode == MODE_UNKNOWN) {
+        return -1;
+    }
+
+    enum mcp2515_mode cur_mode = mcp2515_get_mode(dev);
+
+    if (cur_mode == mode) {
+        return mode;
+    }
+    if (cur_mode == MODE_SLEEP) {
+        mcp2515_wake_up(dev);
+    }
+
+    mcp2515_spi_bitmod(dev, MCP2515_CANCTRL, MCP2515_CANCTRL_REQOP_MASK,
+                       (uint8_t)mode);
+    int cnt = 0;
+
+    while (cur_mode != mode && cnt++ < 2) {
```

Comment:
Magic number?

Reply:
I think the purpose of this loop is to allow a certain number of attempts to set the mode, instead of returning an error straight away if the first attempt fails.

Reply:
Funny this comment had alraedy been made in the original PR hahaa

Reply:
Seems a bit hacky, can you confirm this is needed? And if it needed add a comment?

Reply:
To be honest, I don't really agree that this looks hacky. It's a common technique to make several attempts to perform a certain task.

I cannot confirm it is needed. If I comment it out, it works on my setup. **However**, I can imagine in a real-world application, the first attempts might fail. 
For example: my previous employer had a custom pcb with an AVR and this controller. The ISP lines were connected to the same SPI lines as the mcp was. This often resulted in unreliable initializations of the mcp, because the ISP might still be sending out signals when the AVR had already started running its code. Often it was needed to manually reset the board after programming it. 
This while-structure might help with preventing such issues. 

I'll add a small comment explaining what it is for. (Unless you really object to these lines)